python/gen_requirements.py

```diff
@@ -0,0 +1,391 @@
+    elif piece == "dev":
+      saw_dev = True
+
+    if not isinstance(deps, (list, tuple)):
+      problems.append(f'value for key "{piece}" should be list or tuple, see {deps!r}')
+      continue
+
+    if list(sorted(deps)) != list(deps):
+      problems.append(f"deps for key {piece} must be sorted. Correct order:\n  {list(sorted(deps))!r}")
+
+    piece_deps = set()
+    for d in deps:
+      if CONSTRAINT_REGEX.search(d):
+        problems.append(f"for piece {piece}: dependency {d} should not specify a version. "
+                        "Add it to CONSTRAINTS instead.")
+
+      if d.lower() in piece_deps:
+        problems.append(f'for piece {piece}: dependency {d} listed twice')
+
+      piece_deps.add(d.lower())
+
+  extras_pieces = [k for (k, _) in REQUIREMENTS_BY_PIECE if k not in ("dev", "core")]
+  sorted_extras_pieces = list(sorted(extras_pieces))
+  if sorted_extras_pieces != list(extras_pieces):
+    problems.append('pieces other than "core" and "dev" must appear in alphabetical order'
+                    f"{sorted_extras_pieces}")
+
+  return problems
+
+
+def parse_semver(package, constraint, problems):
+  """Parse a semantic versioning constraint of the form "^X.[.Y[.Z[...]]]]"
+
+  Parameters
+  ----------
+  package : str
+      Name of the package specifying this constraint, for reporting problems.
+  constraint : str
+      The semver constraint. Must start with "^"
+  problems : List[str]
+      A list of strings describing problems that have occurred validating the configuration.
+      Problems encountered while validating constraint are appended to this list.
+
+  Returns
+  -------
+  List[str], int :
+      A 2-tuple. The first element is a list containing an entry for each component in the
+      semver string (components separated by "."). The second element is the index of the
+      component in the list which must not change to meet the semver constraint.
+
+      See "Caret requirements" at https://python-poetry.org/docs/versions/.
+  """
+  m = SEMVER_REGEX.match(constraint[1:])
+  if not m:
+    problems.append(f"{package}: invalid semver constraint {constraint}")
+    return [], 0
+
+  min_ver_parts = list(m.groups())
+
+  # Remove the end
+  while min_ver_parts and min_ver_parts[-1] is None or not min_ver_parts[-1].strip():
+    min_ver_parts.pop()
+
+  for i, p in enumerate(min_ver_parts):
+    try:
+      x = int(p.strip())
+    except ValueError as p:
+      if i == len(parts) - 1 and i > 0:
+        # If a pre-release tag is specified (i.e. 1.2.0pre3), allow.
+        return min_ver_parts, i - 1
+
+      problems.append(
+        f"{package}: semver constraint with non-numeric leading component: {constraint}")
+      return [], 0
+
+    if x:
+      return min_ver_parts, i
+
+  else:
+    # All 0's
+    return min_ver_parts, 0
+
+
+def validate_constraints():
+  problems = []
+
+  if not isinstance(CONSTRAINTS, (list, tuple)):
+    problems.append(f"must be list or tuple, see: {CONSTRAINTS!r}")
```

Comment:
This would be an unnecessary check if you used Python type annotations.

Reply:
you still need to use mypy though right? I don't know that we enforce that in our pre-checkin lint right now